test/statetest/statetest.hpp

```diff
@@ -59,8 +63,66 @@ struct StateTransitionTest
     std::unordered_map<uint64_t, std::string> input_labels;
 };
 
+template <typename T>
+T from_json(const json::json& j) = delete;
+
+template <>
+state::BlockInfo from_json<state::BlockInfo>(const json::json& j);
+
+template <>
+state::State from_json<state::State>(const json::json& j);
+
+template <>
+state::Transaction from_json<state::Transaction>(const json::json& j);
+
 StateTransitionTest load_state_test(const fs::path& test_file);
 
 void run_state_test(const StateTransitionTest& test, evmc::VM& vm);
 
+namespace utils
+{
+inline size_t countl_zero_bytes(uint16_t x)
+{
+    auto upper = uint8_t(x >> 8);
+    auto lower = uint8_t(x & 0xFF);
+    return upper ? 0 : (lower ? 1 : 2);
+}
+
+inline size_t countl_zero_bytes(uint32_t x)
+{
+    auto upper = uint16_t(x >> 16);
+    auto lower = uint16_t(x & 0xFFFF);
+    return upper ? countl_zero_bytes(upper) : 2 + countl_zero_bytes(lower);
+}
+
+inline size_t countl_zero_bytes(uint64_t x)
+{
+    auto upper = uint32_t(x >> 32);
+    auto lower = uint32_t(x & 0xFFFFFFFF);
+    return upper ? countl_zero_bytes(upper) : 4 + countl_zero_bytes(lower);
+}
+
+inline std::string shortest_hex(uint64_t v, bool add_0x = true)
+{
+    const auto prefix = std::string(add_0x ? "0x" : "");
+    size_t num_lz_bytes = countl_zero_bytes(v);
+
+    if (num_lz_bytes == 8)
+        return prefix + "0";
+
+    std::string str;
+
+    for (auto i = num_lz_bytes; i < 8; ++i)
+        str += evmc::hex(static_cast<uint8_t>(v >> 8 * (8 - i - 1)));
```

Comment:
Could also just translate it to string first and then doing a substring to the first non-`0` character.

Reply:
```cpp
inline std::string shortest_hex(uint64_t v, bool add_prefix = true)
{
    if (v == 0) {
        return add_prefix ? "0x0" : "0";
    }

    auto value = hex(v, false);
    // String leading zeroes.
    if (size_t pos = value.find_first_not_of("0"); pos != string::npos) {
        value = value.substr(pos);
    }
    return (add_prefix ? "0x" : "") + value;
}
```